### DISCUSSION TITLE
Testing headless .NET 8 -> .NET 9 update

### DISCUSSION THREAD

**Discussion by Frooxius** [Yellow-Dog-Man/Resonite-Issues#3196] [Pre-release Testing]
Hello everyone!

I've got another test for you & heads up for upcoming update!

## What is being tested
I've updated the headless client from .NET 8 to .NET 9 runtime which has released earlier this month.

This is mostly just a flag change and I don't really expect any breakage to actually happen. However since there's a number of tools in the community, I want to give everyone to give opportunity to prepare for the switch, particularly users hosting events and we might as well use that as opportunity to test this.

https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/3176

## Current build number
2024.11.19.510 

**IMPORTANT:** The build is compatible with current public build as well! You can host sessions with .NET 9 headless already and users will be able to join from the normal non-prelease build.

## How to test
- On Steam, switch to the `prerelease` branch and make sure the build updates before you run it
- Make sure that Steam downloads the new build
- Make sure you have .NET 9 runtime installed
- Run the headless client

### What to test
- Make sure your own tooling works with the version change.
- Host sessions, have people join as you'd normally do

### What NOT to test & report
- Any pre-existing bugs. If you're unsure, switch back to main build and test it again. Report it *only* if it occurs on prerelease (otherwise just make a normal bug report)

## If you find something that got broken
- Create a GitHub issue for a pre-release bug: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/new?assignees=shiftyscales&labels=prerelease&projects=&template=prerelease.yml
- Make sure to include the log
- Please include description of what happened. Be succinct, but exact.

- If you find a roadblocker that's not a bug, but something on your end that would disrupt events and you want us to delay this update, please let us know in this discussion with your reasoning
- Similarly if you for some reason cannot update the tooling in time and we'd disrupt your events, let us know and we can see if we can postpone the release to main

## If you find that stuff works okay
Please report it too and include details on what you generally tested, so we know things are okay. You don't need to make a new issue for this - just a commend in this Discussion is fine.

This lets us know that a bunch of stuff was tested and get better confidence that stuff won't break with release.

## Optional testing
This is not really needed, since it won't change the course of our action in any way, but if you'd like to gather some interesting data, I'd be curious to see how does the performance between .NET 8 headless and .NET 9 headless differ! If you gather some cool data, I'd love to see!

## How long will testing go for?
~~We plan to make the switch to .NET 9 **next Monday**, which is on **November 25th** to give you some time to prepare.~~

~~**UPDATE:** The switch is now postponed to the **Thursday next week**, which is on **November 28th**.~~

**UPDATE 2:** The switch is now postpones to **next Monday**, which is on **December 2nd**.

We can re-evaluate if you hit roadblockers, but please give it a test if you can!

**Comment by bredo228** [comment]
Doing a quick test, the update to .NET 9 seems to be working fine.

**Reply by darbdarb** [reply]
was here decent test, more then a quick test but performance and networking seemed good

**Reply by ModernBalloonie** [reply]
balls

**Reply by ModernBalloonie** [reply]
Seriously though we spawned many charactercontroller balls.

**Reply by bredo228** [reply]
![image](https://github.com/user-attachments/assets/5c6376b6-1483-468e-a491-c7f5fafbb5f7)

**Reply by Frooxius** [reply]
Oh awesome! Did you notice any performance difference compared to .NET 8 headless?

**Reply by bredo228** [reply]
Not noticeably - didn't really have a point of reference to go off.

The headless frame rate remained pretty stable around 60 during most of the testing; unsurprisingly though it struggled with a couple thousand balls.

**Comment by j4lc** [comment]
Added a small migration guide to the wiki (even though this one is really easy): https://wiki.resonite.com/Headless_Server_Software/Migration_Net9

Headless-wise, no bugs to be reported, everything runs fine.

**Reply by Frooxius** [reply]
Awesome, thank you for putting that together!

**Reply by troyBORG** [reply]
It looks like Ubuntu hasn't officially released a .net9 runtime yet in there package manager....
Trying to figure out how to install it manually..

**Reply by j4lc** [reply]
> It looks like Ubuntu hasn't officially released a .net9 runtime yet in there package manager.... Trying to figure out how to install it manually..

You could use the Microsoft repos for the package: https://learn.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#register-the-microsoft-package-repository

**Reply by troyBORG** [reply]
Microsoft's own repo for ubuntu doesn't even have dotnet9 in it...
https://packages.microsoft.com/ubuntu/22.04/prod/pool/main/d/

**Reply by troyBORG** [reply]
Also on your wiki do we need to edit out current config from `-beta headless` to `-beta prerelease` so it downloads the net dotnet9 thing we are testing..  Shouldn't you put you wiki @jae1911

**Reply by j4lc** [reply]
Hmm, then maybe this?

https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual

**Reply by j4lc** [reply]
> Also on your wiki do we need to edit out current config from `-beta headless` to `-beta prerelease` so it downloads the net dotnet9 thing we are testing.. Shouldn't you put you wiki @jae1911

The wiki links back to the top post of this discussion thread which has instructions on how to download it.
By "doesn't requires any configuration changes", I mean that no changes are needed in the `Config.json` file of the headless.

**Reply by troyBORG** [reply]
Opps yeah I didn't mean config.json I meant my `start_headless.sh` script....

**Reply by j4lc** [reply]
> Opps yeah I didn't mean config.json I meant my `start_headless.sh` script....

Yeah, the temporary change to download the test version is explained in the top post of this discussion which is linked on the wiki.

**Reply by troyBORG** [reply]
> Yeah, the temporary change to download the test version is explained in the top post of this discussion which is linked on the wiki.

Not really... It just says "On Steam, switch to the prerelease branch"
That sounds like GUI only explanation.   Not CLI

**Reply by j4lc** [reply]
> Not really... It just says "On Steam, switch to the prerelease branch" That sounds like GUI only explanation. Not CLI

This also applies to the cli as switching branches is done by using `-beta`.

**Reply by gentlecolts** [reply]
My personal docker file is based on arch, and executes  `RUN pacman -Syu --noconfirm steam git lib32-sdl2 base-devel dotnet-runtime lib32-dbus` for dependencies, if that helps anyone.

Doesn't seem like that's pinned to any particular dotnet version, here's to hoping things go smoothly

**Reply by troyBORG** [reply]
I'll mess with it more tonight after work.
One nice thing is mines an LXC container so I can snapshot try stuff if it breaks revert and try other stuff...

**Reply by troyBORG** [reply]
Well even though I was on Ubuntu 22.04.  I ended up using Debian 12 repository to install dotnet9
https://learn.microsoft.com/en-us/dotnet/core/install/linux-debian

Once Ubuntu adds the dotnet9 I'll make it so I'm using the correct repo.
But it worked!
![image](https://github.com/user-attachments/assets/d3483872-23d8-462c-9cfb-5491fd47d735)

**Reply by troyBORG** [reply]
[@Tulip](https://discord.com/channels/1040316820650991766/1154514007479287942/1311031723731189800) mentioned that Ubuntu 24.04 so the backports had it. 
https://launchpad.net/~dotnet/+archive/ubuntu/backports

**Reply by gentlecolts** [reply]
Update: finally got to updating things.  My arch based docker image needed basically no updating.  only change was that arch still has dotnet 8 in the official repos, had to pull 9 from AUR

**Comment by Frooxius** [comment]
**UPDATE:** Due to a request the switch is now postponed to Thursday next week, which is **November 28th**!

**Comment by OrionMoonclaw** [comment]
I updated my headless (hosted on an arm64 device, had to recompile Resonite.dll)

Everything running good so far!

**Comment by Frooxius** [comment]
**UPDATE:** We are postponing the switch until **next Monday on December 2nd**!

**Reply by Alex-2Pi** [reply]
Any specific reason for moving the date again?
I ask because I don't see here or on Discord and information that would suggest it.
Especially since we are supposed to go here for more information.

I'm not saying it's a bad thing that it's moved and just curious if people have asked for a little more time and yet maybe you've discovered a problem that you still want to solve before the change.

**Reply by Frooxius** [reply]
The upcoming event is being hosted on .NET 8 and the stress test was ran with .NET 8, so we're keeping it to eliminate as many variables for things that could go wrong.

**Reply by Alex-2Pi** [reply]
Oh okey, sorry

**Comment by Frooxius** [comment]
The .NET 9 is out now! Thanks everyone for testing!

**Comment by markovg** [comment]
I'm in steamcmd, and 
"app_update 2519830 -beta headless -betapassword <pwd>"
Has no effect ... it says I'm up-to-date, but my build number is 2024.11.19.479 (as per headless output while running) not 2024.11.19.510.  
Since the .NET 9 version is now out, I would expect that as a result of the above command, steam should pickup the new 510 build and update the local install. 

My dotnet version is updated as follows:
sudo add-apt-repository ppa:dotnet/backports
sudo apt install dotnet9

$ dotnet --version
9.0.100

**Reply by darbdarb** [reply]
> I'm in steamcmd, and "app_update 2519830 -beta headless -betapassword " Has no effect ... it says I'm up-to-date, but my build number is 2024.11.19.479 (as per headless output while running) not 2024.11.19.510. Since the .NET 9 version is now out, I would expect that as a result of the above command, steam should pickup the new 510 build and update the local install.
> 
> My dotnet version is updated as follows: sudo add-apt-repository ppa:dotnet/backports sudo apt install dotnet9
> 
> $ dotnet --version 9.0.100

try adding "validate" after
steamcmd.sh +app_update 2519830 -beta headless -betapassword **validate** ${BETA_CODE} +quit

**Reply by markovg** [reply]
Thanks yes this worked! (just "validate" comes after ${BETA_CODE} ), i.e. in steamcmd:
> "app_update 2519830 -beta headless -betapassword ${BETA_CODE} validate"
Then the headless server updated ...
